### PR TITLE
[feat] Support report annotations and add dynamic analyzer related annotations

### PR DESCRIPTION
**Pull request by bruntib**
A plist file may contains a section for report annotations:

```.xml
<key>diagnostics</key>
<array>
  <dict>
    <key>category</key>
    <string>Memory error</string>
    ...
    <dict>
      <key>timestamp</key>
      <string>2000-01-01 10:00</string>
      <key>testsuite</key>
      <string>TS-1</key>
      ...
    </dict>
  </dict>
</array>
```

At the time of creation of this commit there is no analyzer that would
emit a .plist file with such content. But the plan is that sanitizers and
some other dynamic analyzers will provide this information, at least the
report converter could add the timestamp to the .plist files.

The goal of this new feature is to enable storing arbitrary annotations to
the reports. We could have extended the report table with further columns
for each annotation, or we could have joined them in separate tables. Finally
we decided to create a generic "report_annotations" table where the
annotations are collected as (report_id; key; value) tuples. This adds some
complexity when selecting annotations for each report, but this tradeoff is
accepted.

Annotations are untyped labels, but "timestamp" and "testsuite" will be
handled in a specific way in the sense that users can order reports by their
timestamps and can filter reports based on which testsuite resulted them.
This can be useful because in case of dynamic analyzers it is possible that
a report is a consequence of another one. Ordering them by their timestamps
can help in detecting these relationships.

This commit targets the addition of "timestamp" and "testsuite" annotations,
but the implementation is more general than supporting only these two.
This is the reason why the database schema and Thrift API interfaces don't
have the concept of dynamic analysis.

### CODE REVIEW

**Comment by bruntib** [comment]
I'll amend the commit messages, otherwise this PR is ready for review more or less.

File: web/api/report_server.thrift
Comment (Szelethus) [inline web/api/report_server.thrift:586]: Why wouldn't you just create a new function for querying the list of keys? Seems like a poor choice to merge these two functionalities into a single function, not to mention that the name of the function implies that the whole annotation map is going to be returned.

File: web/server/codechecker_server/api/mass_store_run.py
Comment (Szelethus) [inline web/server/codechecker_server/api/mass_store_run.py:988]: Can we do something more friendly? I feel like this validation sacrifices the intelligibility of the error message for being somewhat more general, and it feels like a bad trade.

File: web/server/codechecker_server/api/report_server.py
Comment (Szelethus) [inline web/server/codechecker_server/api/report_server.py:1763]: We needed to stand in front of a white table and look at examples to understand whats happening here. This is the part where we turn rows into columns, right? Could you please write about that a little?

File: web/server/codechecker_server/database/run_db_model.py
Comment (Szelethus) [inline web/server/codechecker_server/database/run_db_model.py:435]: Or maybe some explanation could be here.

**Comment by Szelethus** [comment]
For a major feature, you could add a explanation and reasoning in the summary. Remember, patches like this are most likely to be found during git blame and thoroughly inspected!

> A plist file may contains a section for report annotations:

I mean thats not a lie, but its not the whole story either. There are no tools that emit a plist like that as of writing this comment, so the full story is that **you are adding a new feature**, namely that we will now recognize this pattern in the plist file.

> These annotations are stored in a new database table. "timestamp" and "testsuite" are displayed as a new column in the report table. The goal of these columns is to indicate the exact timestamp of a report. This way the reports of dynamic analyzers can be sorted.

Thats a lot of "how", and not a lot of "why". Why do we need this feature? Why would we want to sort by a timestamp? Why is this more relevant for dynamic analyzers than static analyzers? Why are these two added at the same time? Why are we adding an annotations field to `ReportData` and showing these two traits in it, instead of adding them natively?

Some of the things missing and critical design choices that are hard to understand even the context of this patch. I know the answer to these questions, but you make archeology a pain this way, and I wouldn't like to see this patch merged without a better summary.

I know that "mOaR cOmMeNtS PLZ" is the cheapest and laziest way to review, and I tried to constructively point out where I'm missing some very specific explanations and more importantly, why.

**Review by Szelethus — COMMENTED** [review]
I don't feel comfortable leaving comments on the main part of the patch, namely sqlalchemy, I'll leave that to @vodorok .

**Comment by Szelethus** [comment]
One more thing -- you are doing two different things at one here:
* You are adding annotations to `ReportData`, and the know-how of turning the rows of an annotation table into columns.
* Add timestamp, testsuite, and testcase as new annotations.
It seems to me that these additions would deserve at the very least two pull requests. Again, I wouldn't make you split this right away, but this looks like lot of code, and more importantly a lot of features in a single PR.

File: web/api/js/codechecker-api-node/package.json
Comment (bruntib) [reply web/api/js/codechecker-api-node/package.json:3]: This feature development was started by @vodorok. This PR contains his commit too and there was an increment from 51 to 52. I'm not sure how to exclude that commit without losing his developments.

File: web/api/report_server.thrift
Comment (bruntib) [reply web/api/report_server.thrift:324]: I extended the documentation. However, I'm not sure if "which files to modify" should be included. I think an API documentation is for client applications about what information they can get and how they can get it.

**Review by vodorok — CHANGES_REQUESTED** [review]
I would not mix in refactors in the report filtering login as this feature is not implemented similarly to the old features.

File: web/api/report_server.thrift
Comment (bruntib) [reply web/api/report_server.thrift:586]: I'll remove this whole function and `TestsuiteFilter.vue` completely for now, because these will be needed for the filtering option which we skip for now.

File: web/codechecker_web/shared/version.py
Comment (bruntib) [reply web/codechecker_web/shared/version.py:21]: You did a transition from 51 to 52 in a commit which is the part of this PR. I didn't want to remove your previous developments, but I continued them in a further commit.

File: web/server/codechecker_server/database/run_db_model.py
Comment (bruntib) [reply web/server/codechecker_server/database/run_db_model.py:404]: What is a "mapping construct"?

File: web/tests/functional/dynamic_results/test_dynamic_results.py
Comment (bruntib) [reply web/tests/functional/dynamic_results/test_dynamic_results.py:41]: Sorry, I don't really get this :S

File: web/server/codechecker_server/api/report_server.py
Comment (bruntib) [reply web/server/codechecker_server/api/report_server.py:1805]: I added some docs.

File: web/server/codechecker_server/database/run_db_model.py
Comment (bruntib) [reply web/server/codechecker_server/database/run_db_model.py:404]: Oh, I see now. I saw this documentation earlier, but it seems to be like an alternative way of defining columns. In this commit I followed the original methodology of defining ORM tables.

File: web/server/codechecker_server/api/report_server.py
Comment (bruntib) [reply web/server/codechecker_server/api/report_server.py]: I added some docs about the reasons.

File: web/api/js/codechecker-api-node/package.json
Comment (vodorok) [reply web/api/js/codechecker-api-node/package.json:3]: If you want, you can drop my original commits, but If I remember correctly you can regenerate the thrift stubs on the same version number if you wanted. after you made your modifications. Are there any implications of the double-step bump apart from "looking weird"?

File: web/server/codechecker_server/api/mass_store_run.py
Comment (Szelethus) [reply web/server/codechecker_server/api/mass_store_run.py]: I still think this should emit a friendlier error message, even if it means some less elegant code.

File: web/api/report_server.thrift
Comment (Szelethus) [reply web/api/report_server.thrift]: Since this is a string-string map, it is critical for the server to validate whether the message is proper. I think at the very least, lets bring attention to the importance of validation in `__validate_and_add_report_annotations`.

**Review by Szelethus — CHANGES_REQUESTED** [review]
I appreciate some of the new documentation, but the patch (and the summary, and the commit message) still fails on describing the design philosophy. There is a lot of emphasis on _how_ stuff works, and very little if any about _why_ it works that way, and why not in another.

**Review by vodorok — APPROVED** [review]
lets wait for tests but lgtm

**Pull request merged**
